### PR TITLE
Replicate stack gan

### PR DESCRIPTION
**Pull request by Devin-Taylor**
* image normalisation
* added residual connections
* abstracted repeating layers into custom layers
* general clean-up

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -69,6 +69,10 @@ def parse_arguments(args_to_parse):
         '--conditional-emb-size', type=int, help='The number of elements in the conditiona embedding',
         default=default_settings['conditional_emb_size']
     )
+    training.add_argument(
+        '--aug', type=bool, help='Flag if augmentation must be applied to training data',
+        default=default_settings['augment']
+    )
```
Comment (alecokas) [inline main.py:75]: Is this not a data argument rather than training?

File: main.py [hunk 2]
```diff
@@ -155,7 +159,8 @@ def main(args):
             batch_size=args.batch_size,
             save_location=results_dir,
             num_embeddings=default_settings['num_embeddings'],
-            num_samples=default_settings['num_samples']
+            num_samples=default_settings['num_samples'],
+            augment=default_settings['augment']
```
Comment (alecokas) [inline main.py:163]: Does this mean that the `--aug` CLI parameter does not actually do anything?

File: models/discriminators.py [hunk 3]
```diff
@@ -139,12 +115,10 @@ def loss(self, predictions_on_real, predictions_on_fake):
                 predictions_on_real : Tensor
                 predictions_on_fake : Tensor
         """
-        real_loss = tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.ones_like(predictions_on_real), logits=predictions_on_real)
+        real_loss = tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.fill(predictions_on_real.shape, 0.9), logits=predictions_on_real)
```
Comment (alecokas) [inline models/discriminators.py:118]: I don't really understand this change. Why use 0.9 rather than 1?

File: models/generators.py [hunk 4]
```diff
@@ -104,73 +107,48 @@ def __init__(self, img_size, num_latent_dims, kernel_size, num_filters,
...
+        self.deconv_block_3 = DeconvBlock(128, self.w_init, self.bn_init, activation=True)
+
+        self.deconv2d_4 = Conv2DTranspose(num_output_channels, kernel_size=(4, 4), strides=(2, 2), padding='same', kernel_initializer=self.w_init)
+        self.conv2d_4 = Conv2D(filters=num_output_channels, kernel_size=(3, 3), strides=(1, 1), padding='same', kernel_initializer=self.w_init)
 
     @tf.function
     def call(self, embedding, noise):
         smoothed_embedding, mean, log_sigma = self.conditional_augmentation(embedding)
-        noisy_embedding = tf.concat([smoothed_embedding, noise], 1)
+        noisy_embedding = tf.concat([noise, smoothed_embedding], 1)
```
Comment (alecokas) [inline models/generators.py:131]: Does the order here matter? I'm OK with the change, but I'm just not sure if I'm missing something

File: models/generators.py [hunk 5]
```diff
@@ -180,7 +158,7 @@ def loss(self, predictions_on_fake, mean, log_sigma):
         """ Only calculate the loss based on the discriminator
             predictions for the images generated by this model.
         """
-        kl_coeff = 1
+        kl_coeff = 2
```
Comment (alecokas) [inline models/generators.py:161]: Is this change in the paper? Just checking because I don't understand

File: models/layers.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        self.activation = activation
+        self.conv2d = Conv2D(filters=filters, kernel_size=kernel_size, strides=strides, padding=padding, kernel_initializer=w_init)
+        self.bn = BatchNormalization(gamma_initializer=bn_init)
+        self.leaky_relu = LeakyReLU(alpha=0.2)
+
+    def call(self, x):
+        x = self.conv2d(x)
+        x = self.bn(x)
+        if self.activation:
+            x = self.leaky_relu(x)
```
Comment (alecokas) [inline models/layers.py:64]: Very nice :+1:

File: main.py (hunk 2 above)
Comment (Devin-Taylor) [reply main.py:163]: yeah, just going to remove the argument

File: models/discriminators.py (hunk 3 above)
Comment (Devin-Taylor) [reply models/discriminators.py:118]: Took it from a repo, I think I will change it back actually it was more just for testing.

File: models/generators.py (hunk 4 above)
Comment (Devin-Taylor) [reply models/generators.py:131]: no different, just OCD haha

File: models/generators.py (hunk 5 above)
Comment (Devin-Taylor) [reply models/generators.py:161]: yeah, paper
